src/commands/giveaway.rs

```diff
@@ -0,0 +1,650 @@
+            b.label("Previous page");
+            b.disabled(offset - 10 < 0)
+        });
+        r.create_button(|b| {
+            b.style(ButtonStyle::Secondary);
+            b.custom_id("GIVEAWAY_list_next");
+            b.label("Next page");
+            b.disabled(offset + 10 >= giveaways)
+        })
+    });
+    c
+}
+
+async fn generate_list_embeds(db: &Database, offset: i64) -> Vec<CreateEmbed> {
+    let giveaways = db.get_n_giveaways_with_offset(10, offset).await.unwrap();
+    let mut giveaway_winners = Vec::with_capacity(giveaways.len());
+
+    for g in giveaways.iter() {
+        let winners = db.get_giveaway_winners(g.id).await.unwrap();
+        giveaway_winners.push(winners);
+    }
+
+    let mut embeds: Vec<CreateEmbed> = Vec::new();
+    for (g, winners) in giveaways.iter().zip(giveaway_winners.iter()) {
+        let mut e = CreateEmbed { 0: HashMap::new() };
+        let winner_string = winners
+            .iter()
+            .map(|x| format!("<@{}>", x.user_id.to_string()))
+            .collect::<Vec<String>>()
+            .join(", ");
+
+        e.title(format!("Giveaway #{}", g.id));
+        e.description(format!(
+            "**Prize**: {}\n**Winners**: {}\n**End time**: <t:{}:R>",
+            g.prize,
+            if g.completed {
+                winner_string
+            } else {
+                format!("Max {}", g.max_winners)
+            },
+            g.end_time.timestamp()
+        ));
+        embeds.push(e);
+    }
+    embeds
+}
+
+async fn get_reacters(
+    http: &Http,
+    message: &Message,
+    reaction: ReactionType,
+) -> Result<Vec<User>, anyhow::Error> {
+    let mut total_users: Vec<User> = Vec::new();
+    loop {
+        match message
+            .reaction_users(
+                &http,
+                reaction.clone(),
+                Some(100),
+                total_users.last().map(|x| x.id),
+            )
+            .await
+        {
+            Ok(mut users) => {
+                if users.is_empty() {
+                    return Ok(total_users);
+                }
+                total_users.append(&mut users);
+            }
+            Err(e) => {
+                return Err(e.into());
+            }
+        }
+    }
+}
+
+async fn roll_winners(candidates: &Vec<User>, max_winners: i64) -> Vec<User> {
+    candidates
+        .choose_multiple(&mut rand::thread_rng(), max_winners as usize)
+        .map(|x| x.to_owned())
+        .collect()
+}
+
+async fn reroll_giveaway(
+    http: &Http,
+    db: &Database,
+    giveaway: &Giveaway,
+    reaction: ReactionType,
+) -> Result<(), anyhow::Error> {
+    let mut message = http
+        .get_message(giveaway.channel_id, giveaway.message_id)
+        .await?;
+    let candidates = get_reacters(&http, &message, reaction.clone())
+        .await
+        .map(|v| v.into_iter().filter(|x| !x.bot).collect::<Vec<User>>())?;
+    let winners = roll_winners(&candidates, giveaway.max_winners).await;
```

Comment:
In case of a reroll, do we want the same member (pre-reroll-winner) to be able to win again?

Reply:
Perhaps a "allow past winners" boolean should be added to the reroll command?

Reply:
Yes, that would work.